packages/statemanager/src/cache/storage.ts

```diff
@@ -0,0 +1,374 @@
+    if (this._lruCache) {
+      storageMap = this._lruCache!.get(addressHex)
+    } else {
+      storageMap = this._orderedMapCache!.getElementByKey(addressHex)
+    }
+    this._stats.reads += 1
+    if (storageMap) {
+      this._stats.hits += 1
+      return storageMap.getElementByKey(keyHex)
+    }
+  }
+
+  /**
+   * Marks storage key for address as deleted in cache.
+   * @param address - Address
+   * @param key - Storage key
+   */
+  del(address: Address, key: Uint8Array): void {
+    const addressHex = bytesToHex(address.bytes)
+    const keyHex = bytesToHex(key)
+    this._saveCachePreState(addressHex, keyHex)
+    if (this.DEBUG) {
+      this._debug(`Delete storage for ${addressHex}: ${keyHex}`)
+    }
+    if (this._lruCache) {
+      let storageMap = this._lruCache!.get(addressHex)
+      if (!storageMap) {
+        storageMap = new OrderedMap()
+      }
+      storageMap.setElement(keyHex, hexStringToBytes('80'))
+      this._lruCache!.set(addressHex, storageMap)
+    } else {
+      let storageMap = this._orderedMapCache!.getElementByKey(addressHex)
+      if (!storageMap) {
+        storageMap = new OrderedMap()
+      }
+      storageMap.setElement(keyHex, hexStringToBytes('80'))
+      this._orderedMapCache!.setElement(addressHex, storageMap)
+    }
+
+    this._stats.dels += 1
+  }
+
+  /**
+   * Deletes all storage slots for address from the cache
+   * @param address
+   */
+  clearContractStorage(address: Address): void {
+    const addressHex = bytesToHex(address.bytes)
+    if (this._lruCache) {
+      this._lruCache!.set(addressHex, new OrderedMap())
+    } else {
+      this._orderedMapCache!.setElement(addressHex, new OrderedMap())
+    }
+  }
+
+  /**
+   * Flushes cache by returning storage slots that have been modified
+   * or deleted and resetting the diff cache (at checkpoint height).
+   */
+  async flush(): Promise<[string, string, Uint8Array | undefined][]> {
+    if (this.DEBUG) {
+      this._debug(`Flushing cache on checkpoint ${this._checkpoints}`)
+    }
+
+    const diffMap = this._diffCache[this._checkpoints]!
+    const it = diffMap.begin()
+
+    const items: [string, string, Uint8Array | undefined][] = []
+
+    while (!it.equals(diffMap.end())) {
+      const addressHex = it.pointer[0]
+      const diffStorageMap = it.pointer[1]
+      let storageMap: StorageCacheMap | undefined
+      if (this._lruCache) {
+        storageMap = this._lruCache!.get(addressHex)
+      } else {
+        storageMap = this._orderedMapCache!.getElementByKey(addressHex)
+      }
+
+      if (storageMap !== undefined) {
+        const itDiffStorage = diffStorageMap.begin()
+        while (!itDiffStorage.equals(diffStorageMap.end())) {
+          const keyHex = itDiffStorage.pointer[0]
+          const value = storageMap.getElementByKey(keyHex)
+          items.push([addressHex, keyHex, value])
+          itDiffStorage.next()
+        }
+      } else {
+        throw new Error('Inconsistent cache state')
```

Comment:
This is just an internal error similar to the one Jochem mentioned above signaling an implementation error if thrown since normally `storageMap` should be defined at that point.

Reply:
It is, though maybe we could make the message a little more descriptive like `storageMap should be defined at this point` or something so I or other team members who don't yet grasp all of the cache inner workings? :grimacing:

Reply:
Yes, makes sense, have updated.